rig-core/src/providers/gemini/completion.rs

```diff
@@ -0,0 +1,649 @@
+                    .tools
+                    .into_iter()
+                    .map(Tool::from)
+                    .collect(),
+            ),
+            tool_config: None,
+            system_instruction: Some(Content {
+                parts: vec![Part {
+                    text: "system".to_string(),
+                }],
+                role: Some("system".to_string()),
+            }),
+        };
+
+        tracing::info!("Request: {:?}", request);
+
+        let response = self
+            .client
+            .post(&format!("/v1beta/models/{}:generateContent", self.model))
+            .json(&request)
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<ApiResponse<GenerateContentResponse>>()
+            .await?;
+
+        match response {
+            ApiResponse::Ok(response) => Ok(response.try_into()?),
+            ApiResponse::Err(err) => Err(CompletionError::ResponseError(err.message)),
+        }
+    }
+}
+
+impl From<completion::ToolDefinition> for Tool {
+    fn from(tool: completion::ToolDefinition) -> Self {
+        Self {
+            function_declaration: FunctionDeclaration {
+                name: tool.name,
+                description: tool.description,
+                parameters: None, // tool.parameters, TODO: Map Gemini
+            },
+            code_execution: None,
+        }
+    }
+}
+
+impl TryFrom<GenerateContentResponse> for completion::CompletionResponse<GenerateContentResponse> {
+    type Error = CompletionError;
+
+    fn try_from(response: GenerateContentResponse) -> Result<Self, Self::Error> {
+        match response.candidates.as_slice() {
+            [ContentCandidate { content, .. }, ..] => Ok(completion::CompletionResponse {
+                choice: completion::ModelChoice::Message(
+                    content.parts.first().unwrap().text.clone(),
+                ),
+                raw_response: response,
+            }),
+            _ => Err(CompletionError::ResponseError(
+                "No candidates found in response".into(),
+            )),
+        }
+    }
+}
+
+impl TryFrom<serde_json::Value> for GenerationConfig {
+    type Error = CompletionError;
+
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        let mut config = GenerationConfig {
+            temperature: None,
+            max_output_tokens: None,
+            stop_sequences: None,
+            response_mime_type: None,
+            response_schema: None,
+            candidate_count: None,
+            top_p: None,
+            top_k: None,
+            presence_penalty: None,
+            frequency_penalty: None,
+            response_logprobs: None,
+            logprobs: None,
+        };
+
+        fn unexpected_type_error(field: &str) -> CompletionError {
+            CompletionError::ResponseError(format!("Unexpected type for field '{}'", field))
+        }
+
+        if let Some(obj) = value.as_object() {
```

Comment:
This check fails if `additional_params` is not set, even though the `GenerationConfig` is not required to query the API.